hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -97,126 +280,355 @@ impl<'t> YkPTBlockIterator<'t> {
+                            ObjLoc::OtherObjOrUnknown(opt_vaddr) => match opt_vaddr {
+                                Some(vaddr) => vaddr,
+                                None => unreachable!(),
+                            },
+                        }
+                    };
+                    dis.set_ip(u64::try_from(ret_vaddr).unwrap());
+                    reposition = true;
+                }
+                iced_x86::FlowControl::IndirectBranch | iced_x86::FlowControl::IndirectCall => {
+                    self.seek_tip()?;
+                    let vaddr = match self.cur_loc {
+                        ObjLoc::MainObj(off) => self.off_to_vaddr(&PHDR_MAIN_OBJ, off)?,
+                        ObjLoc::OtherObjOrUnknown(opt_vaddr) => match opt_vaddr {
+                            Some(vaddr) => vaddr,
+                            None => unreachable!(),
+                        },
+                    };
+
+                    if inst.flow_control() == iced_x86::FlowControl::IndirectCall {
+                        if inst.next_ip() as usize == vaddr {
+                            todo!("zero length call");
+                        }
+                        debug_assert!(!inst.is_call_far());
+                        self.comprets
+                            .push(CompRetAddr::VAddr(usize::try_from(inst.next_ip()).unwrap()));
                     }
-                    SuccessorKind::Dynamic => {
-                        // We can only know the successor via a TIP update in a packet. Fall
-                        // through to `self.seek_tip()`.
-                        self.cur_loc = ObjLoc::OtherObjOrUnknown;
+
+                    dis.set_ip(u64::try_from(vaddr).unwrap());
+                    reposition = true;
+                }
+                iced_x86::FlowControl::ConditionalBranch => {
+                    // Ensure we have TNT decisions buffered.
+                    if self.tnts.is_empty() {
+                        self.seek_tnt()?;
                     }
+                    // unwrap() cannot fail as the above code ensures we have decisions buffered.
+                    if self.tnts.pop_front().unwrap() {
+                        dis.set_ip(self.branch_target_vaddr(&inst));
+                        reposition = true;
+                    }
+                }
+                iced_x86::FlowControl::UnconditionalBranch => {
+                    dis.set_ip(self.branch_target_vaddr(&inst));
+                    reposition = true;
                 }
-            } else {
-                // In the absence of blockmap info, we cannot know the successor block.
-                self.cur_loc = ObjLoc::OtherObjOrUnknown;
-                return Ok(Some(Block::unknown()));
+                iced_x86::FlowControl::Call => {
+                    if inst.code() == iced_x86::Code::Syscall {
+                        // Do nothing. We assume the kernel isn't traced.
```

Comment:
Does a syscall (seem to) immediately return?

Reply:
It makes a `TIP.PGD` packet (tracing disabled) upon entry to the kernel, and a `TIP.PGE` (tracing enabled) packet upon reentry to userspace. We handle these already.

There is the option to trace the kernel, but as we've discussed, that'd be nonsense.

Reply:
Can we clarify the comment to something like "We've turned off kernel tracing elsewhere, so syscalls are TIP.PGD immediately followed by TIP.PGE"?

Reply:
Fixed in 0fa48b4f824eb454261fe5378557e0d6581a4149